### PR TITLE
Jason/request courses

### PR DESCRIPTION
**Pull request by JasonBaoZ**
Created backend functionality and rudimentary front-end functionality for requesting new courses once student is done with their existing courses. Allows student to select courses they want to request. Allows admin to approve or reject those requests and see the courses they want to request.

screen shot 2017-04-14 at 7 26 16 pm
screen shot 2017-04-14 at 7 27 23 pm
@lbkchen @amillman

### CODE REVIEW

**Comment by charlesx2013** [comment]
sik css bruh

**Comment by lbkchen** [comment]
This is a @charlesx2013 level page styling

**Comment by charlesx2013** [comment]
ouch

File: app/controllers/api/students/requests_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class Api::Students::RequestsController < Api::Students::BaseController
+  load_and_authorize_resource
+
+  def create
+    @request.student_id = current_user.id
```
Comment (lbkchen) [inline app/controllers/api/students/requests_controller.rb]: @charlesx2013 is this proper

File: 0 [hunk 2]
```diff
@@ -0,0 +1,2 @@
+      invoke  active_record
+      create    db/migrate/20170322010437_requests.rb
```
Comment (amillman) [inline 0]: what is this

File: app/assets/javascripts/shared/routes.js [hunk 3]
```diff
@@ -65,6 +69,8 @@ class RailsRoutes {
   static courseEditPath(id)           { return `/courses/${id}/edit` }
   static adminProfilePath(id)         { return `/admins/${ReactRoutes.getId(id, 'id')}/profile`}
   static codeCsvListPath()            { return `/admins/code_csvs`}
+  //is this needed?
+  //static courseRequestPath()          { return `/students/requests`}
```
Comment (amillman) [inline app/assets/javascripts/shared/routes.js]: if you dont need this then remove the comment

File: app/controllers/api/admins/requests_controller.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+class Api::Admins::RequestsController < Api::Admins::BaseController
+  load_and_authorize_resource
+
+  def update
+    puts @request
```
Comment (amillman) [inline app/controllers/api/admins/requests_controller.rb]: remove this

File: app/models/request.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+#  student_id :integer
+#  state      :integer          default(0)
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+
+class Request < ActiveRecord::Base
+  validates :student_id, presence: true
+
+  has_many :course_requests
```
Comment (amillman) [inline app/models/request.rb]: if you want to use `course_requests` as a join table, you should use `has_many :courses through: :course_requests`. then you can do something like `request.courses`

Comment (amillman) [inline app/models/request.rb]: something that might be useful is to also add this `has_many :through` relation on the course too, if someone wanted to query the requests for a given course

File: app/controllers/api/admins/code_csvs_controller.rb [hunk 6]
```diff
@@ -1,5 +1,5 @@
 class Api::Admins::CodeCsvsController < Api::Admins::BaseController
-  load_and_authorize_resource
+  load_resource
```
Comment (lbkchen) [inline app/controllers/api/admins/code_csvs_controller.rb]: Going to revert this

**Pull request merged**